cartographer_grpc/client_server_test.cc

```diff
@@ -42,22 +75,88 @@ class ClientServerTest : public ::testing::Test {
       return MAP_BUILDER_SERVER)text";
     auto map_builder_server_parameters =
         cartographer::mapping::test::ResolveLuaParameters(kMapBuilderServerLua);
-    auto map_builder_server_options =
+    map_builder_server_options_ =
         CreateMapBuilderServerOptions(map_builder_server_parameters.get());
-    auto map_builder =
-        cartographer::common::make_unique<cartographer::mapping::MapBuilder>(
-            map_builder_server_options.map_builder_options());
+    const std::string kTrajectoryBuilderLua = R"text(
+      include "trajectory_builder.lua"
+      return TRAJECTORY_BUILDER)text";
+    auto trajectory_builder_parameters =
+        cartographer::mapping::test::ResolveLuaParameters(
+            kTrajectoryBuilderLua);
+    trajectory_builder_options_ =
+        cartographer::mapping::CreateTrajectoryBuilderOptions(
+            trajectory_builder_parameters.get());
+  }
+
+  void InitializeRealServer() {
+    auto map_builder = cartographer::common::make_unique<MapBuilder>(
+        map_builder_server_options_.map_builder_options());
     server_ = cartographer::common::make_unique<MapBuilderServer>(
-        map_builder_server_options, std::move(map_builder));
+        map_builder_server_options_, std::move(map_builder));
     EXPECT_TRUE(server_ != nullptr);
   }
 
+  void InitializeServerWithMockMapBuilder() {
+    auto mock_map_builder = cartographer::common::make_unique<MockMapBuilder>();
+    mock_map_builder_ = mock_map_builder.get();
+    server_ = cartographer::common::make_unique<MapBuilderServer>(
+        map_builder_server_options_, std::move(mock_map_builder));
+    EXPECT_TRUE(server_ != nullptr);
+  }
+
+  void InitializeStub() {
+    stub_ = cartographer::common::make_unique<mapping::MapBuilderStub>(
+        map_builder_server_options_.server_address());
+    EXPECT_TRUE(stub_ != nullptr);
+  }
+
+  proto::MapBuilderServerOptions map_builder_server_options_;
+  MockMapBuilder* mock_map_builder_;
+  cartographer::mapping::proto::TrajectoryBuilderOptions
+      trajectory_builder_options_;
   std::unique_ptr<MapBuilderServer> server_;
   std::unique_ptr<mapping::MapBuilderStub> stub_;
 };
 
 TEST_F(ClientServerTest, StartAndStopServer) {
+  InitializeRealServer();
+  server_->Start();
+  server_->Shutdown();
+}
+
+TEST_F(ClientServerTest, StartAndStop) {
+  InitializeRealServer();
+  server_->Start();
+  InitializeStub();
```

Comment:
Did you check initializing the sub actually fails if you don't start the server? Just to make sure this test actually tests something other than what StartAndStopServer above already tests.

Reply:
Right, initializing the stub does not require the serve to run, so I remove this test.